### DISCUSSION TITLE
OOM in ReferenceCountedOpenSslEngine due to suppressed exceptions accumulating in pendingException

### DISCUSSION THREAD

**Discussion by tkowalcz** [netty/netty#13349] [Q&A] [answered]
`ReferenceCountedOpenSslEngine` has a `pendingException` field, it's purpose seems to be to store exceptions, do some cleanup and rethrow them later at a more convenient time. There is also code that appends subsequent exceptions to `pendingException` as `suppressed`:

```java
ThrowableUtil.addSuppressed(pendingException, exception);
```

In my particular case this led to OOM error where `ReferenceCountedOpenSslEngine` occupied almost 8gb of heap with 12 million of suppressed exceptions.

Screenshot from Eclipse MAT:
<img width="1046" alt="image" src="https://user-images.githubusercontent.com/309611/233486130-9ad57b1b-9d62-417b-96ed-460575faffe6.png">

Netty version where this occurred is `4.1.53.Final`. It does not look like handling these exceptions is different compared to master. We also used `netty-tcnative-boringssl-static` in version `2.0.46.Final`.

What might be interesting is that the underlying SSL error is `TOO_MANY_KEY_UPDATES`. I found [this report similar](https://lists.apache.org/list.html?user@cassandra.apache.org) but without OOM. Here is the BoringSSL code reporting this error: https://github.com/google/boringssl/blob/bcecc7d834fc44ad257b2f23f88e1cf597ab2736/ssl/tls13_both.cc#L677.

I have trouble reproducing `TOO_MANY_KEY_UPDATES` error. I will appreciate any pointers on how to trigger that, then I'll try to make a reproducer based on master branch.

**Comment by normanmaurer** [comment, marked as answer]
@tkowalcz converted to an issue... Also see https://github.com/netty/netty/pull/13351
